modules/pre-commit.nix

```diff
@@ -315,7 +329,31 @@ in
                 echo 1>&2 "    3. add .pre-commit-config.yaml to .gitignore"
               else
                 ln -s ${configFile} .pre-commit-config.yaml
-                pre-commit install
+                # Remove any previously installed hooks (since pre-commit iteslf has no idempotent design)
+                hooks="pre-commit pre-merge-commit pre-push prepare-commit-msg commit-msg post-checkout post-commit"
+                for hook in $hooks; do
+                  pre-commit uninstall -t $hook
+                done
+                # Add hooks for configured stages (only) ...
+                if [ ! -z "${concatStringsSep " " cfg.default_stages}" ]; then
+                  for stage in ${concatStringsSep " " cfg.default_stages}; do
+                    case $stage in
+                      commit | merge-commit | push)
+                        stage="pre-"$stage
+                        pre-commit install -t $stage
+                        ;;
+                      prepare-commit-msg | commit-msg | post-checkout | post-commit)
+                        stage="pre-"$stage
+                        pre-commit install -t $stage
+                        ;;
```

Comment:
I see what you're trying to do here, but it's broken. Let's stick closely to git and pre-commit's terminology. It's easier to learn for those who are familiar with it and it's more future proof (any new hooks will be ignored)

Reply:
We have to acknowledge it's a fundamental flaw of pre-commit:

1. On one hand it tries to be as declarative as possible in version controlled `.pre-commit-config.yml`
2. On the other hand any _sound_ distributed configuration on the developer machine is out of the reach of the devops team (without employing an auxiliary task runner, such as `make` or `just`).

Since a `shell.nix` is expected to produce a reproducible environment, this flaw is fundamentally incompatible with `shell.nix` philosophy.

Hence, since we don't want to betray ourselves (`shell.nix`), fixing this non-converging design of pre-commit is a moral imperative.

The side effects introduced by pre-commit into the local `.git/hooks/...` _must_ be controlled.

I'd rather prefer to do help message parsing to make this future proof, than to let those side effects rot in there to eternity.

Reply:
I filed an upstream feature request: https://github.com/pre-commit/pre-commit/issues/1541

Reply:
> moral imperative

I'm taking this with a shovel of salt, considering that it's just some tech in the end. I wouldn't be so sure if someone outside the functional programming bubble (like pre-commit's maintainer and the python community) appreciates this kind of 'grand' language about concepts that are unnecessary from a pragmatic point of view.

Anyway, this seems to be about the broader topic of improving declarative hook installation, whereas my comment was originally intended about the `"pre-"` prefixing. I totally failed to make that clear. The idea is that we should treat the names of commit hooks atomically rather than making them the subject of string manipulation. The reason to do so is that it'd be easier for users to switch between pre-commit-hooks.nix configuration, git manpages and pre-commit.com documentation when the terminology is aligned one-to-one.

Reply:
Is there any reason not to simplify the whole `case ... esac` statement into just `pre-commit install -t $stage`? The second clause is equal to the first (probably by mistake?) but I don't think we really need to do anything special here anyway.

Reply:
by mistake, thanks!

In `pre-commit`, stages `commit | merge-commit | push` are named different from hooks `pre-commit | pre-merge-commit | pre-push`. It's very confusing, but who are we to tell... :wink:

Reply:
force pushed

Reply:
Ok, we'll just deal with that. No problem. One more comment about the default case though, but I'll make that a separate thread.